### DISCUSSION TITLE
gridpatchdataset->apply transformation?

### DISCUSSION THREAD

**Discussion by janetkok** [Project-MONAI/MONAI#1473] [Q&A] [answered]
Is it possible to apply transformation for gridpatchdataset ? 
and how to use dataloader to load gridpatchdataset cause gridpatchdataset has no len()

Thanks

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help share some comments here?
I think you can show more usage of `GridPatchDataset`.

Thanks in advance.

**Reply by wyli** [reply]
perhaps we could now add support of patch-level transforms for the grid dataset? if @ericspod agrees I'll create a feature request

**Reply by ericspod** [reply]
The linked solution uses a previous data set as input so permits only whole image transforms. Let's add that feature request in case there is a need for patch-level operations.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Maybe you can share more details about the usage of `GridPatchDataset` in research work?
Thanks in advance.

**Comment by ericspod** [comment, marked as answer]
There is an example of how to use it in the [Nifti Read Example tutorial notebook](https://github.com/Project-MONAI/tutorials/blob/master/modules/nifti_read_example.ipynb), does this answer your questions?

**Reply by nveeturi2** [reply]
I tried the example # 6 from above link to create a data loader which yields patches in regular grid order from loaded images, but ran into issues. Could you please help?
MONAI version: 0.4.0+150.g401ea29
Numpy version: 1.19.5
Pytorch version: 1.7.1
**Details below:**
---> 29 im, seg = first(loader)
TypeError: Caught TypeError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 198, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 28, in fetch
    data.append(next(self.dataset_iter))
  File "/home/ubuntu/anaconda3/lib/python3.7/site-packages/monai/data/grid_dataset.py", line 161, in __iter__
    for patch, slices, *_ in self.patch_iter(image):  # patch_iter to yield at least 2 items: patch, coords
TypeError: 'tuple' object is not callable
PS:
I was able to successfully use GridPatchDataset with PatchIter, but I need to crop image and mask image both as part of a transformation using Compose(i.e. instead of RandSpatialCropSamples with limited samples, I want to use all the patches created using GridPatchDataset) but ran into an issue with DataLoader.

**Reply by ericspod** [reply]
I think the `patch_iter` argument you're passing when constructing `GridPatchDataset` isn't correct. This has to be a callable which will return an iterator producing patches from the input image, such as `PatchIter`. It looks like what you've passed in doesn't meet this criteria.

**Reply by nveeturi2** [reply]
Thanks for the reply. In section 6 example, we dont pass patch_iter. The implementation using patchIter is working fine. Here's the code I tried:
```
imtrans = Compose([LoadImage(image_only=True),
                   ScaleIntensity(), AddChannel(), ToTensor()])

segtrans = Compose([LoadImage(image_only=True), AddChannel(), ToTensor()])

ds = ArrayDataset(images, imtrans, segs, segtrans)
ds = GridPatchDataset(ds, (64, 64, 64))

loader = torch.utils.data.DataLoader(
    ds, batch_size=10, num_workers=2, pin_memory=torch.cuda.is_available()
)
im, seg = first(loader)
print(im.shape, seg.shape)
```

**Reply by ericspod** [reply]
It looks like the `GridPatchDataset` has changed since I mentioned the Nifti reading tutorial. I think `GridPatchDataset(ds, PatchIter((64, 64, 64)))` would work. We should also fix the example of course. @rijobro

**Reply by nveeturi2** [reply]
Thanks, I will try this and get back.
